### PR TITLE
feat: GPU support

### LINKED ISSUE #127: Feature request: GPU support
## Feature

Enable GPU acceleration for embedding generation.

## Motivation

My CPU fan is noisy.

### PR DESCRIPTION
**Pull request by mickey-mikey**
Closes #127

## What this changes

Adds an optional `RAG_DEVICE` environment variable.

- When unset or `cpu`, the embedder calls `pipeline(..., { device: 'cpu' })` (current default).
- When set to any other string, that string is passed through to transformers.js.
- If `pipeline()` rejects the device, the embedder throws `EmbeddingError` on first use and the underlying transformers.js error message reaches the caller.

No allowlist is maintained in this repo; the list of accepted device names is whatever the linked `@huggingface/transformers` version accepts.

## CI

Matrix expanded to `[ubuntu-latest, windows-latest] × [cpu, webgpu]`.

The `webgpu` cells install a software Vulkan adapter (SwiftShader on Windows, Lavapipe on Linux) and set `WGPU_BACKEND=vulkan` plus `VK_ICD_FILENAMES` so Dawn binds to it. CI runners have no GPU. Native EPs (`dml`/`cuda`/`coreml`) require hardware that GitHub-hosted runners don't provide.

Lint, format, and type-check were moved into a separate `static-checks` job so they run once per OS instead of once per (OS, device) cell.

## Test plan

- [ ] `static-checks` and both `test` matrix dimensions pass on Ubuntu and Windows
- [ ] `RAG_DEVICE` unset → server logs `device=cpu`
- [ ] `RAG_DEVICE=<unsupported>` → first embed throws `EmbeddingError` with the transformers.js message

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -444,7 +445,7 @@ PDF, DOCX, TXT, Markdown, and HTML (via `ingest_data`). Not yet: Excel, PowerPoi
 Yes, but you must delete your database and re-ingest all documents. Different models produce incompatible vector dimensions.
 
 **GPU acceleration?**
-Transformers.js runs on CPU. GPU support is experimental. CPU performance is adequate for most use cases.
+Opt-in via `RAG_DEVICE`. Verified: `webgpu` offloads compute to the GPU on Windows with a real D3D12-capable adapter (`nvidia-smi` confirms the Node process on the GPU). Linux and macOS WebGPU paths exist but are not verified in this repo. For maximum throughput pick the platform-native EP: `dml` (Windows), `cuda` (Linux x64 NVIDIA), `coreml` (macOS). Default is `cpu`. The server throws if the requested device fails — set `RAG_DEVICE=cpu` to revert.
```
Comment (shinpr) [inline README.md]: "For maximum throughput pick the platform-native EP" overgeneralizes. On macOS arm64 / Apple Silicon, ingest of a 65-page PDF with the default model was ~3s on `cpu`, ~5s on `webgpu`, ~44s on `coreml`. One workload on one host — not enough to call these devices useless, but enough that "platform-native = throughput win" doesn't hold beyond the Windows + D3D12 + NVIDIA case verified in the PR. The `coreml` number in particular says platform-native isn't free.

The accepted-values row on line 323 also mixes user-recommended devices with advanced pass-through values. `auto` and `gpu` work but reintroduce upstream auto-selection, which contradicts the PR's "no auto-detection" stance. `wasm` passes the PR's validation and then throws at `pipeline()` init in at least the macOS arm64 / Node 22 backend tested here (`Unsupported device: "wasm". Should be one of: coreml, webgpu, cpu.`).

Suggested replacement:

Configuration table (line 323):

```
| `RAG_DEVICE` | — | `cpu` | Advanced: execution device for embedding inference. Keep `cpu` unless you are benchmarking your own hardware. Common values: `webgpu`, `dml`, `cuda`, `coreml`. |
```

FAQ entry (lines 447–448):

```
**GPU acceleration?**
`RAG_DEVICE` is an advanced opt-in for embedding inference. The default `cpu` setting is recommended for most users. Non-CPU devices may improve throughput or reduce CPU load on some hardware, models, and workloads, but they can also be slower or unsupported. Benchmark your own workload before using them permanently.

Common values:
- `webgpu`: cross-platform GPU path where WebGPU is available
- `dml`: Windows DirectML
- `cuda`: Linux x64 with NVIDIA/CUDA support
- `coreml`: macOS Core ML

If a selected device fails, set `RAG_DEVICE=cpu`.
```

File: src/embedder/index.ts [hunk 2]
```diff
@@ -1,6 +1,30 @@
 // Embedder implementation with Transformers.js
 
-import { env, pipeline } from '@huggingface/transformers'
+import { type DeviceType, env, pipeline } from '@huggingface/transformers'
+
+const VALID_DEVICES: readonly DeviceType[] = [
```
Comment (shinpr) [inline src/embedder/index.ts]: The allowlist is the upstream Transformers.js `DeviceType` union rather than a curated set for this deployment. `auto` and `gpu` are auto-selection keywords — they don't match "no auto-detection". `wasm` and `webnn`, `webnn-npu`, `webnn-gpu`, `webnn-cpu` fail at `pipeline()` init in at least the macOS arm64 / Node 22 backend tested here — `RAG_DEVICE=wasm` throws `Unsupported device: "wasm". Should be one of: coreml, webgpu, cpu.` Trim to `cpu, webgpu, dml, cuda, coreml`.

File: src/embedder/index.ts [hunk 3]
```diff
@@ -1,6 +1,30 @@
...
+  'cuda',
+  'dml',
+  'coreml',
+  'webnn',
+  'webnn-npu',
+  'webnn-gpu',
+  'webnn-cpu',
+] as const
+
+function resolveDevice(): DeviceType {
```
Comment (shinpr) [inline src/embedder/index.ts]: This runs at lazy init, not at server startup. `RAGServer.initialize()` only initializes `vectorStore`, so `RAG_DEVICE=bogus` survives startup and the `status` tool, and dies on the first tool call. Lift the validation into `server-main.ts` env parsing — add a `parseDevice(process.env['RAG_DEVICE'])` alongside the other parsers around line 92 of that file, and put the resolved value on `RAGServerConfig.device`. Once it lives there, the `Configuration:` log on line 146 of `server-main.ts` surfaces it for free, so "what device am I actually on?" doesn't have to wait for the first embed.

File: src/embedder/index.ts [hunk 4]
```diff
@@ -66,20 +90,25 @@ export class Embedder {
...
         dtype: 'fp32',
+        device,
       })
-      console.error('Embedder: Model loaded successfully')
+      console.error(`Embedder: Model loaded successfully (device=${device})`)
     } catch (error) {
       throw new EmbeddingError(
-        `Failed to initialize Embedder: ${(error as Error).message}`,
+        `Failed to initialize Embedder on device "${device}": ${(error as Error).message}\n\n` +
+          'Set RAG_DEVICE=cpu to force CPU. Other values: gpu (auto per platform), dml, cuda, coreml, webgpu, wasm, auto.',
```
Comment (shinpr) [inline src/embedder/index.ts]: Hand-written valid-values list; already drifts from `VALID_DEVICES` (no `webnn*`). If the array on line 5 is trimmed, interpolate `${VALID_DEVICES.join(', ')}` or point at the README rather than restate it.

File: src/server/index.ts [hunk 5]
```diff
@@ -741,4 +741,12 @@ export class RAGServer {
     await this.server.connect(transport)
     console.error('RAGServer running on stdio transport')
   }
+
+  /**
+   * Stop the server and release resources
+   */
+  async close(): Promise<void> {
```
Comment (shinpr) [inline src/server/index.ts:752]: Two gaps:

1. `this.embedder` isn't closed. transformers.js `Pipeline.dispose()` (`@huggingface/transformers/src/pipelines/_base.js:111`) releases the underlying ONNX session, and on real GPU backends the device's GPU resources/VRAM. Add `Embedder.close()` that calls `(this.model as { dispose?: () => Promise<void> }).dispose?.()` and clears `this.model`, then `await this.embedder.close()` here.
2. The MCP SDK exposes `Server.close()`, but `RAGServer.close()` doesn't call it. As written, "Stop the server and release resources" only releases LanceDB.

File: src/vectordb/index.ts [hunk 6]
```diff
@@ -476,4 +476,17 @@ export class VectorStore {
       throw new DatabaseError('Failed to get status', error as Error)
     }
   }
+
+  /**
+   * Close the database connection
+   */
+  async close(): Promise<void> {
```
Comment (shinpr) [inline src/vectordb/index.ts:483]: `ftsEnabled` isn't reset. Only matters if anyone reuses the instance after close → initialize, which no caller does today. Low priority.

File: vitest.config.mjs [hunk 7]
```diff
@@ -10,8 +10,8 @@ export default defineConfig({
     globals: true,
     environment: 'node',
     // Process management improvements
-    testTimeout: 10000,        // 10 second timeout
-    hookTimeout: 10000,        // Hook processing timeout 10 seconds
+    testTimeout: 60000,        // 60 second timeout
+    hookTimeout: 120000,       // Hook processing timeout 120 seconds
```
Comment (shinpr) [inline vitest.config.mjs]: `hookTimeout` 10s → 120s without a comment. Plausibly for the 90MB cold-cache model download in `beforeAll`, but the diff doesn't say so. Add a one-line rationale. Worth revisiting after `Embedder.dispose()` lands, since teardowns should then be cheap.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -36,6 +46,34 @@ jobs:
...
+    - name: Install Vulkan Dependencies (Ubuntu, webgpu only)
+      if: matrix.device == 'webgpu' && runner.os == 'Linux'
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y libvulkan1 vulkan-tools libwayland-client0 mesa-vulkan-drivers
+
+    - name: Verify Vulkan (webgpu only)
+      if: matrix.device == 'webgpu'
+      run: |
+        vulkaninfo --summary || echo "vulkaninfo failed"
```
Comment (shinpr) [inline .github/workflows/ci.yml]: `vulkaninfo --summary || echo "vulkaninfo failed"` makes Vulkan setup failure non-fatal, so a broken setup is reported later from `pnpm test`, farther from the failing setup step. Drop the `|| echo`.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -36,6 +46,34 @@ jobs:
...
+
+    # The integration tests use Xenova/all-MiniLM-L6-v2 via @huggingface/transformers,
+    # which is downloaded on first use (~90MB). Caching it across runs avoids a
+    # network round-trip on every CI build.
+    - name: Cache HuggingFace model
```
Comment (shinpr) [inline .github/workflows/ci.yml]: Cache key is `runner.os`-scoped, not device-scoped, so both `cpu` and `webgpu` cells on the same OS can race to reserve/save the same key. Not fatal, but concurrent cells will produce a cache reservation/save warning on the loser. If you want cleaner logs, gate the cache save on one device cell or split it into a setup job.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -36,6 +46,34 @@ jobs:
...
+      uses: actions/cache@0057852bfaa89a56745cba8c7296529d2fc39830 # v4
+      with:
+        path: tmp/models
+        key: hf-model-${{ runner.os }}-Xenova-all-MiniLM-L6-v2
+
     - name: Lint check
```
Comment (shinpr) [inline .github/workflows/ci.yml]: Lint, format:check, type-check don't depend on `RAG_DEVICE`. Per-OS makes sense (onnxruntime-node binaries, path separators); per-device doubles them for no signal. Move into a separate job that doesn't enumerate the device axis.

File: src/__tests__/e2e/html-workflow.e2e.test.ts [hunk 11]
```diff
@@ -34,6 +34,7 @@ describe('HTML Workflow E2E', () => {
   }, 120000) // 2 minutes for model download
 
   afterAll(async () => {
+    await server.close()
```
Comment (shinpr) [inline src/__tests__/e2e/html-workflow.e2e.test.ts:37]: This `await server.close()` was added here, but the same pattern is missing from `src/__tests__/e2e/rag-workflow.e2e.test.ts` at lines 76, 129, and 219 (the `finally` blocks of all three `it` cases). The LanceDB Windows file-handle pattern needs the same fix there.

**Review by shinpr — CHANGES_REQUESTED** [review]
Requesting changes. The direction is right, but the implementation and docs overstate what `RAG_DEVICE` guarantees: validation happens too late, the app-level allowlist is broader than what the runtime supports, lifecycle cleanup skips the embedder pipeline, and the README frames non-CPU devices as maximum-throughput choices without benchmark coverage.

Five items below should land before merge. The rest are second-pass.

## Blockers

1. **Device validation runs at first embed, not at server startup.** The PR description says "init throws", and the README says "The server throws if the requested device fails". `RAGServer.initialize()` only touches `vectorStore`, so `RAG_DEVICE=bogus` survives startup and the `status` tool, and fails on the first `query_documents`/`ingest_file`. Move the string check into `server-main.ts` env parsing and pass the resolved value through `RAGServerConfig` to `EmbedderConfig`. Lazy model load can stay.
2. **The whitelist contradicts "no auto-detection".** `auto` and `gpu` are upstream auto-selection keywords. `wasm` and `webnn*` pass app validation but fail at `pipeline()` init in at least the macOS arm64 / Node 22 backend tested here — `RAG_DEVICE=wasm` throws `Unsupported device: "wasm". Should be one of: coreml, webgpu, cpu.` That turns an up-front, helpful rejection into a downstream opaque one. Trim to `cpu, webgpu, dml, cuda, coreml`.
3. **No dispose path for the embedder pipeline.** `Pipeline.dispose()` exists in transformers.js and releases the underlying ONNX session; on real GPU backends it also releases the device's GPU resources/VRAM. Add `Embedder.close()` that calls `pipeline.dispose()`, then call it from `RAGServer.close()`. While there, also call `this.server.close()` — the MCP SDK exposes it but `RAGServer.close()` doesn't currently call it, so `close()` only releases LanceDB.
4. **No unit tests for `resolveDevice()`, and the wrapped-error paths in `ensureInitialized` / `embedBatch` are also untested.** Boundary cases: invalid, empty, whitespace-only, case-variants, unset. Easier to write once validation lives in `server-main.ts`.
5. **The README's `RAG_DEVICE` framing overgeneralizes.** "For maximum throughput pick the platform-native EP" — on macOS arm64 / Apple Silicon, `coreml` ingest of a 65-page PDF with the default model took ~44s vs ~3s on `cpu` (~14× slower). One workload on one machine, but enough to show the platform-native recommendation doesn't transfer beyond the Windows + D3D12 + NVIDIA case the PR verified. The accepted-values row on README:323 also mixes user-recommended devices with advanced pass-through values: `auto` and `gpu` reintroduce upstream auto-selection (contradicting the PR's "no auto-detection" stance), and `wasm` fails at runtime in at least the macOS arm64 / Node 22 backend tested here. Reframe `RAG_DEVICE` as an advanced opt-in and drop "maximum throughput" wording without benchmark coverage. Concrete replacement text on `README.md:448`.

## Second pass

- CI: `vulkaninfo --summary || echo ...` swallows Vulkan setup failure. Drop the `|| echo`.
- CI: lint / format:check / type-check don't depend on `RAG_DEVICE` and currently run twice per OS. Split into a once-per-OS job.
- CI: HF model cache key is OS-scoped, not device-scoped, so both device cells per OS race to save the same key. Cosmetic.
- CI / README: the "exercises the same code path" framing for the webgpu CI cells is true at the WebGPU dispatch layer, not at GPU execution — SwiftShader/Lavapipe is a software substitute. Worth tightening the wording.
- `vitest.config.mjs` hookTimeout 10s → 120s is plausibly justified by the 90MB cold-cache model download in `beforeAll`, but the diff carries no rationale. Add a one-line comment; revisit after dispose lands.
- `src/__tests__/e2e/rag-workflow.e2e.test.ts` doesn't call `ragServer.close()` in its `finally` blocks; other test files in this PR were updated for the same handle pattern.
- `VectorStore.close()` leaves `ftsEnabled` set. Only matters if a caller reuses the instance across close → initialize, which no caller does today.
- `src/embedder/index.ts:138` — `ensureInitialized()` re-wraps `EmbeddingError`; early-rethrow `EmbeddingError` so invalid `RAG_DEVICE` stays a config error instead of gaining unrelated network/disk/cache guidance.
- `src/embedder/index.ts:160` — empty-text validation runs after `ensureInitialized()`; hoist it above initialization to avoid loading the model before rejecting `""`.
- `src/embedder/index.ts:217` — `embedBatch()` has the same `EmbeddingError` re-wrap pattern as `ensureInitialized()`.

File: src/embedder/index.ts (hunk 2 above)
Comment (mickey-mikey) [reply src/embedder/index.ts]: Curation is inflexible. Prefer to punt the decision list to upstream. Upstream decides the allowlist and documentation can point to the source for guidance. Not ideal UX, but the target audience is developers so flexibility matters more than polish here.

**Review by shinpr — APPROVED** [review]
Thanks for the follow-up and fixes. CI and local checks are green on my side, and the remaining concern is addressed by the pass-through direction you clarified.
